src/subtype.c

```diff
@@ -3060,13 +3065,15 @@ static jl_value_t *intersect_varargs(jl_vararg_t *vmx, jl_vararg_t *vmy, ssize_t
 {
     // Vararg: covariant in first parameter, invariant in second
     jl_value_t *xp1=jl_unwrap_vararg(vmx), *xp2=jl_unwrap_vararg_num(vmx),
-                *yp1=jl_unwrap_vararg(vmy), *yp2=jl_unwrap_vararg_num(vmy);
+               *yp1=jl_unwrap_vararg(vmy), *yp2=jl_unwrap_vararg_num(vmy);
+    jl_value_t *i2=NULL, *ii = intersect(xp1, yp1, e, param==0 ? 1 : param);
+    if (ii == jl_bottom_type)
+        return jl_bottom_type;
     // in Vararg{T1} <: Vararg{T2}, need to check subtype twice to
     // simulate the possibility of multiple arguments, which is needed
     // to implement the diagonal rule correctly.
-    if (intersect(xp1, yp1, e, param==0 ? 1 : param) == jl_bottom_type)
-        return jl_bottom_type;
-    jl_value_t *i2=NULL, *ii = intersect(xp1, yp1, e, 1);
+    if (!((xp2 && jl_is_long(xp2) && jl_unbox_long(xp2) == 1) || (yp2 && jl_is_long(yp2) && jl_unbox_long(yp2) == 1)))
+        ii = intersect(xp1, yp1, e, 1);
```

Comment:
Doesn't this imply that xp2 cannot get narrowed later to be assigned the value `1` or `0`, or is that handled elsewhere?

Reply:
IIUC, we never do that if `Vararg` has fixed length? (We usually unwrap them and fill the `parameters`, thus their lengths are unchangeable.)

Reply:
Ah yes, this check doesn’t make much sense as `Vararg` might have been consumed. Perhaps we should let `intersect_tuple` handle these cases.